js/actions/AppActions.js

```diff
@@ -35,65 +36,70 @@ import { browserHistory } from 'react-router';
-    bcrypt.hash(password, salt, (err, hash) => {
-      // Something wrong while hashing
-      if (err) {
-        requestFailed({
-          type: 'failed'
-        });
-        return;
-      }
-      // Use auth.js to fake a request
-      auth.login(username, hash, (success, err) => {
-        // When the request is finished, hide the loading indicator
-        dispatch(sendingRequest(false));
-        dispatch(setAuthState(success));
-        if (success === true) {
-          // If the login worked, forward the user to the dashboard and clear the form
-          forwardTo('/dashboard');
-          dispatch(changeForm({
-            username: "",
-            password: ""
-          }));
-        } else {
-          requestFailed(err);
-        }
-      });
-    });
-  }
+	return (dispatch) => {
+		// Show the loading indicator, hide the last error
+		dispatch(sendingRequest(true));
+		// If no username or password was specified, throw a field-missing error
+		if (anyElementsEmpty({ username, password })) {
+			dispatch(setErrorMessage(errorMessages.FIELD_MISSING));
+			dispatch(sendingRequest(false));
+			return;
+		}
+		// Generate salt for password encryption
+		const salt = genSalt(username);
+		// Encrypt password
+		bcrypt.hash(password, salt, (err, hash) => {
+			// Something wrong while hashing
+			if (err) {
+				dispatch(setErrorMessage(errorMessages.GENERAL_ERROR));
+				return;
+			}
+			// Use auth.js to fake a request
+			auth.login(username, hash, (success, err) => {
+				// When the request is finished, hide the loading indicator
+				dispatch(sendingRequest(false));
+				dispatch(setAuthState(success));
+				if (success === true) {
+					// If the login worked, forward the user to the dashboard and clear the form
+					forwardTo('/dashboard');
+					dispatch(changeForm({
+						username: "",
+						password: ""
+					}));
+				} else {
+					switch (err.type) {
+						case 'user-doesnt-exist':
+							dispatch(setErrorMessage(errorMessages.USER_NOT_FOUND));
+							return;
+						case 'password-wrong':
+							dispatch(setErrorMessage(errorMessages.WRONG_PASSWORD));
+							return;
+						default:
+							dispatch(setErrorMessage(errorMessages.GENERAL_ERROR));
+							return;
+					}
+				}
+			});
+		});
+	}
 }
 
 /**
  * Logs the current user out
  */
 export function logout() {
-  return (dispatch) => {
-    dispatch(sendingRequest(true));
-    auth.logout((success, err) => {
-      if (success === true) {
-        dispatch(sendingRequest(false));
-        dispatch(setAuthState(false));
-        browserHistory.replace(null, '/');
-      } else {
-        requestFailed(err);
-      }
-    });
-  }
+	return (dispatch) => {
+		dispatch(sendingRequest(true));
+		auth.logout((success, err) => {
+			if (success === true) {
+				dispatch(sendingRequest(false));
+				dispatch(setAuthState(false));
+				browserHistory.replace(null, '/');
+			} else {
+				dispatch(setErrorMessage(errorMessages.GENERAL_ERROR));
+			}
+		});
+	}
```

Comment:
Please fix your editor, you (unnecessarily) changed all the whitespace in this file 😉